libopenage/gui/game_control_link.cpp

```diff
@@ -0,0 +1,334 @@
+// Copyright 2015-2016 the openage authors. See copying.md for legal info.
+
+#include "game_control_link.h"
+
+#include <algorithm>
+
+#include <QtQml>
+
+#include "../engine.h"
+#include "engine_link.h"
+#include "game_main_link.h"
+
+namespace openage {
+namespace gui {
+
+namespace {
+const int registration_mode = qmlRegisterUncreatableType<OutputModeLink>("com.openage", 1, 0, "OutputMode", "OutputMode is an abstract interface for the concrete modes like EditorMode or ActionMode.");
+const int registration_create = qmlRegisterType<CreateModeLink>("com.openage", 1, 0, "CreateMode");
+const int registration_action = qmlRegisterType<ActionModeLink>("com.openage", 1, 0, "ActionMode");
+const int registration_editor = qmlRegisterType<EditorModeLink>("com.openage", 1, 0, "EditorMode");
+const int registration = qmlRegisterType<GameControlLink>("com.openage", 1, 0, "GameControl");
+}
+
+OutputModeLink::OutputModeLink(QObject *parent)
+	:
+	GuiItemQObject{parent},
+	QQmlParserStatus{},
+	GuiItemInterface<OutputModeLink>{} {
+}
+
+OutputModeLink::~OutputModeLink() {
+}
+
+QString OutputModeLink::get_name() const {
+	return this->name;
+}
+
+QStringList OutputModeLink::get_binds() const {
+	return this->binds;
+}
+
+void OutputModeLink::on_announced(const std::string &name, const std::vector<std::string>& binds) {
+	auto new_name = QString::fromStdString(name);
+
+	if (this->name != new_name) {
+		this->name = new_name;
+		emit name_changed();
+	}
+
+	QStringList new_binds;
+	std::transform(std::begin(binds), std::end(binds), std::back_inserter(new_binds), [] (const std::string &s) {return QString::fromStdString(s);});
+
+	if (this->binds != new_binds) {
+		this->binds = new_binds;
+		emit binds_changed();
+	}
+}
+
+void OutputModeLink::classBegin() {
+}
+
+void OutputModeLink::on_core_adopted() {
+	QObject::connect(&unwrap(this)->gui_signals, &OutputModeSignals::announced, this, &OutputModeLink::on_announced);
+}
+
+void OutputModeLink::componentComplete() {
+	static auto f = [] (OutputMode *_this) {
+		_this->announce();
+	};
+	this->i(f);
+}
+
+CreateModeLink::CreateModeLink(QObject *parent)
+	:
+	Inherits{parent} {
+	Q_UNUSED(registration_create);
+}
+
+CreateModeLink::~CreateModeLink() {
+}
+
+ActionModeLink::ActionModeLink(QObject *parent)
+	:
+	Inherits{parent},
+	food{},
+	wood{},
+	gold{},
+	stone{} {
```

Comment:
Again, we shouldn't fix the resource types :) Think of awesome mods adding whatever you can imagine. Oil, silicon, ...